### ISSUE TITLE
RFC: social cards

### ISSUE THREAD

**Issue by slorber** [facebook/docusaurus#2968]
Issue for idea already expressed here: https://github.com/facebook/docusaurus/issues/2267#issuecomment-639597335

## 💥 Proposal

# Dynamic Social Cards

You might have noticed that some people / companies have an automated way to generate social cards that adapt to the underlying content (generating an image from query string params)

I think this would cool if by default, all Docusaurus docs/blogs would generate custom social cards, if the user did not define any custom social image.

This way, all Docusaurus links pasted to a social media would have a beautiful image. 

By default we could even include a discrete Docusaurus + Slash logo for growth hacking 🤪 

Vercel has a nice project to do this easily: https://github.com/vercel/og-image

We could fork this project, host our own variant (can be free on Vercel), and use it for D2 website, + allow other sites to use it as well.

Curretly, Docusaurus can have a default social card, but it's static and does not adapt to the content::

![image](https://user-images.githubusercontent.com/749374/83898560-5fd4db00-a757-11ea-8021-7442ed6fa4c5.png)

Here's some examples of dynamic/automated social cards:

![image](https://user-images.githubusercontent.com/749374/83897173-872aa880-a755-11ea-9cf5-dba1ecd7264a.png)

![image](https://user-images.githubusercontent.com/749374/83897249-a3c6e080-a755-11ea-875e-caec89b0c5ff.png)

![image](https://user-images.githubusercontent.com/749374/83897287-b214fc80-a755-11ea-8aa1-29eadf578d8f.png)

![image](https://user-images.githubusercontent.com/749374/83897394-d5d84280-a755-11ea-8f19-ff284d8ebc50.png)

---

Notes, there are many ways to handle these social cards. 
- We could generate the cards as a build step, using a headless chromium instance in the CIs that support it.
- We could host lambdas to generate the images remotely
- We could use saas services like Cloudinary...

**Comment by slorber** [comment]
@JoelMarcey 

>  I think this is a good idea and I would be interesting in having this implemented. As far as how to implement it, I have no strong feelings. I am generally in favor of utilizing other solutions for specific features if we believe that the cost of implementing it ourselves would be so much greater. That said, we lose a little bit of control of our destiny by doing that. I am definitely curious which way others think we should go here.

What I think about potential implementations:

### Generating the images at build time 

Would require some kind of headless browser in the build process. I think it's likely to complicate the build process, make it slower, less portable, and potentially exceeding free quotas of popular free CIs. I think it's not the best solution.

Seen this being done here with Gasby:
https://andrewingram.net/posts/automatic-social-cards-with-gatsby/

### Using a SaaS

Some people are using a SaaS like Cloudinary to generate these cards, so they don't have to host any infrastructure.

For example: 
https://www.learnwithjason.dev/blog/auto-generate-social-image/ 


### Self hosted

We could create our own solution and host it.

Other solutions are possible, but I mean forking https://github.com/vercel/og-image as a `docusaurus-images` project, and hosting it on Vercel for free (or negotiate some kind of free open-source usage with them).

We would make the service flexible enough so that it works great for most of the sites, yet enabling users to self host an alternative service if our default service does not suit their needs.


---

What I think we should do:
- Create a `docusaurus-images` (can be a fork of vercel/og-image)
- Make it flexible enough for our need (querystring should enable to customize the social card image, enable/disable a Docusaurus stamp, show docs version...)
- Host it on Vercel, making it available  by default to all without having to do anything
- Expose a configuration function that takes params of the page, and return the og image url, (which by default would use the service we host)

This way, we don't complicate the build by default, yet we enable users to craft their own social image urls. They could self-host their own fork of vercel/og-images, use external services like Cloudinary, or even add the image generation as a build step using a docusaurus postBuild plugin.

**Comment by FocalChord** [comment]
hi! me and @Drewbi will be working on this issue :)

**Comment by slorber** [comment]
Great!

@anshulrgoyal  also suggested we could simply use an image processing lib like `sharp` to create the card at build time. That's simpler than using a headless chrome, and that's worth creating a POC to see how far we can go with this approach.

**Comment by anshulrgoyal** [comment]
I was suggesting creating some template for cards and then allowing the user to provide custom stuff like text, icons, logo, images cards and we can have templates that can read mark down and extract stuff from it like for blog we can extract image and title, author and create image from it.

**Comment by slorber** [comment]
Yes, in the end we should be able to add a meta og image tag in generated pages + an image in the build folder. We also need to think about where we plug this image file generation step, we don't have anything like that in D2 afaik. Also need to think about how to make it generic so that it can work for docs, blog, pages... what's the url of such image etc... Not so simple :)

**Comment by Drewbi** [comment]
I'd personally prefer customizing the existing [og-image](https://github.com/vercel/og-image) project as many of the features are very relevant to what we are trying to achieve and all that would be required in terms of modification would be altering the html/css that gets generated internally by the function.

In order to integrate the process locally at build time we would have to create and store an image for every page and either keep track of updates or regenerate the images on the next build. While possible, at this point I feel the API would lend itself quite nicely to the job and would allow for a light weight and easy implementation. It would also allow for users to swap out endpoints if they want like @slorber mentioned.

The customization of fields would be as easy as including or leaving out fields from the query string, eg. `og-docusaurus.now.sh/main%20text?logo=""&author=""&sitetitle=""&docv="2.13.4"` Any of these fields could be left out and it could be handled by the API.

Either way, if we agree on some desirable fields, I can start creating designs/mockups 🎨

**Comment by slorber** [comment]
Some random thoughts.

I think it should be integrated at the theme classic level. Other themes would be responsible for adding themselves the og image system they want (if they do).

I think the theme should allow a function to generate the og image url for each page. It would take some data we provide as parameter

```js
const config { 
  themeConfig: {
    createOGImage: ({data,defaultImage}) => {
      if (data.type === "myCustomDocType" {
        return "https://cloudinary.com/someThing?"+serialize({title: data: title});
      }
      return defaultImage; // default image would handle doc/blog/page...
    }
  }
}
```

Eventually, we could allow the user to generate himself the images at build time using an async function?

```js
const config { 
  themeConfig: {
    createOGImage: async (data) => {
        const relativeFilePath = path.join("og_images",random()+".jpg");
        const filePath = path.join(__dirname,"build","og_images",random()+".jpg");
        await writeFile(filePathgenerateOgImageWithSharp(data));
        return "https://myDocusaurusSite/"+baseUrl + relativeFilePath;
    }
  }
}
```

This function couldn't be called from React directly (as the browser does not have access to nodejs apis like sharp/fs etc), so I think the blog/docs plugin should call it with the blog/doc metadata, and inject it into the theme components as props. (BlogPostPage/DocItem)

---

Not totally sure about this api, feedbacks welcome. We should probably make the system generic enough so that it can be used as well for custom pages (+ the upcoming markdown pages: https://github.com/facebook/docusaurus/pull/2947). 



--- 

@Drewbi I think it's safe to say that we need at least two types of social cards:
- docs: probably including the title, version...
- blog: probably including the title, author, author picture, tags...

In all cases I think including the site logo is nice, and an opt-out Docusaurus logo for growth hacking 🤪

**Comment by anshulrgoyal** [comment]
@slorber I was thinking of adding it to some place like StaticSiteGenerator plugin. We can allow the user to modify options and pass a custom function that will be envoked inside ejs, with all the page meta data

**Comment by anshulrgoyal** [comment]
Some site bots don't render javascript

**Comment by slorber** [comment]
> @slorber I was thinking of adding it to some place like StaticSiteGenerator plugin. We can allow the user to modify options and pass a custom function that will be envoked inside ejs, with all the page meta data

That could be nice to see a poc of this, even if there are lots of things hardcoded in the PR, to see if it can fit the usecases. 
But this means we would perhaps need additional metadata to be able to inject back all the required data to the fn (for example, the type of doc "blog"/"doc"."page"/other...)

I think it's ok to use custom meta elements anyway so it shouldn't be a big deal, + we could make a Helmet wrapper to help the user provide us the data needed for custom pages.

> Some site bots don't render javascript

I don't understand 😅

**Comment by Drewbi** [comment]
Made some initial experiments, am still not completely satisfied with any so will continue experimenting but figure I'd put them out there if anyone has suggestions for things to try or comments about which ones work best.

https://xd.adobe.com/view/5f228e2f-3db9-4755-91be-2b6f95e04de2-0a71/

**Comment by slorber** [comment]
That looks like a nice start, great.

For versioned docs, I'd say we should make the version smaller and closer to the bottom.
And for the Docusaurus icon, I think many people don't know it so it's rather write something like "<icon> - published with Docusaurus" in very small.  But would have to see it to be sure :p

**Comment by sarthakkundra** [comment]
@slorber I'd like to take this as on now. 
CC: @lex111

**Comment by slorber** [comment]
Yes @sarthakkundra thanks 👍 

Do you have good design skills to make some proposals for the social cards? 
We should rather have a different card for each kind of content type, yet all the cards should be consistent with each others imho.

Also worth looking at existing PR draft code from former MLH (just for inspiration, does not mean we have to keep that existing code)

**Comment by sarthakkundra** [comment]
@slorber I have terrible design skills :/ but there is an Adobe XD wireframe link in this thread which I can use? I think that'll go with OG-Vercel too as we can change Vercel's logo with D2's and probably play around with the font.

I had a look at the PR. Will take inspiration from it once I've setup the Vercel API to return Images / SVGs based on our query string.

Let me know if you have an implementation in mind or want to add something to this approach.

**Comment by slorber** [comment]
I don't think you need to master a design tool, as long as you can make a pretty html/css card that's already nice

I don't have much more API ideas that what is written here (it's been a long time I didn't think about this problem).
The best is probably to give a try based on my former comments and propose an implementation, and we'll try to find its shortcomings once we have a POC?

**Comment by sarthakkundra** [comment]
Sounds good!

**Comment by anshulrgoyal** [comment]
@sarthakkundra I also built a POC but it explored an alternate approach, it might worth looking into.

**Comment by sarthakkundra** [comment]
@slorber Made some mockups. Let me know the feedback, will modify og-Vercel accordingly to create our Image as a service provider which we can later host. [Soocial Cards Mockup](https://whimsical.com/rfc-social-cards-849LNmwho81FTY94taJKZw)

**Comment by slorber** [comment]
Thanks, that looks nice. 
We probably want more padding at the top.

**Comment by sarthakkundra** [comment]
@slorber I have hosted the OG image service. Here's a [preview](https://og-image-delta-six.vercel.app/Design%20Principles.png?theme=light&imageType=Docs&md=1&fontSize=100px&images=https%3A%2F%2Fv2.docusaurus.io%2Fimg%2Fdocusaurus_keytar.svg). Let me know your thoughts. Also if required I can write a small documentation for the URL query parameters.

**Comment by slorber** [comment]
yes that looks nice @sarthakkundra 

That would be helpful to also work on the UI/form here https://og-image-delta-six.vercel.app/
This way users will more easily be able to play with all the possible variations.

It's probably better to make the text slightly larger At the same time we should handle nicely docs/blogs with both small/long titles. We can let the user customize the size, and by default choose a good size according to the length of the content?
(We can probably truncate and limit to 2 lines)

**Comment by sarthakkundra** [comment]
@slorber Yes. This was just a POC will make the UI match with our preferences.

**Comment by slorber** [comment]
Problem:
- themeConfig.socialCardFn is complicated to achieve, as configFn can't be executed on client-side theme code

What I suggest for now is:
- Have themeConfig.socialCardService.url, default to https://og-image-delta-six.vercel.app/
- Have themeConfig.socialCardService.defaultParams: the default params added to the social card service querystring
- Create a @theme/SocialCard component, with some meaningful props to construct the appropriate image url with querystring and put it in the page with <Head>
- Use SocialCard in DocPage and BlogPage with relevant props

It's not perfect but it lets:
- User provide his own social card service  url that he could customize by forking the default one
- Swizzle @theme/SocialCard if he wants to have full control over querystring params and final url construction

**Comment by yiliansource** [comment]
This sounds very exciting, may I ask what the progress with this is?

**Comment by slorber** [comment]
No progress has been made on this

**Comment by yiliansource** [comment]
I see. I really like the design mockups by @Drewbi, they're a great start. What I struggle to understand is the following:

The latest stance seems to be to have a service running to generate the social cards dynamically, without having to build them. But doesn't this reliability on an "external" service and the constant generation on their end, whenever the preview should be displayed, become sort of a liability?

From what I can see, there were already plans to do this during build time, using some kind of image processing library. This could definitely be restricted to production builds, to avoid development builds with `docusaurus start` taking ages - social card previews aren't of much use there anyway. Performing the build locally provides the luxury of having the static assets in place reliably and loading them fast. Styling could be done via basic, templated HTML (similar to swizzled components?), which is rendered to JPGs during the static build. Of course, we could even make this toggle-able between static generation (local URLs) and on-the-fly service generation (external URLs).

Is this issue up for grabs? Could be fun to give this a shot, if we're on the same page with the ideas!

**Comment by slorber** [comment]
@yiliansource nobody is working on it currently, and I won't either, so if you want to make a POC you can.

I'd like this to be enabled by default for all sites. And some sites have a lot of docs, so if we build the snapshots locally, we'd rather have a good caching layer for generated images or it will be time-consuming for large sites by default. 

A remote service to render images is a SPOF that we have to maintain, but I think we could provide that to the community: it wouldn't be too complicated to run on Netlify lambdas with an immutable caching strategy.

We should probably offer to generate assets both locally and remotely, and can think about the default behavior later.

**Comment by yiliansource** [comment]
@slorber i'll see if I can craft something up. I'll be honest, I haven't really been using Netlify/Vercel at all yet, but better starting late with it than never.

The argument of quantity is fair - I guess we'll go for the lambda service first then. I'll try to stir something up similar to https://github.com/vercel/og-image, deployed with Vercel so we can test it. Once we're happy with the generator we can look into integrating into Docusaurus.

I don't assume there's a standard or desired URL/query pattern that the service should provide?

**Comment by slorber** [comment]
> I don't assume there's a standard or desired URL/query pattern that the service should provide?

We have to design the url/query API that the service will provide, and also pass all the meaningful params in a custom async callback provided by the user in case he wants to use another service for generating images.

For now having a POC of this is already nice, we can think later about what exactly the api/params should be, but I can think of title, page type, icon...

**Comment by yiliansource** [comment]
Quick update: I managed to patch together an image generation service at https://github.com/yiliansource/docusaurus-og-image, however there are some issues with my current setup when deployed to Vercel, which I'll have to look into.

![image](https://user-images.githubusercontent.com/41469451/123509378-69d4bf00-d675-11eb-8501-a8f48ffbb8fa.png)

**Comment by martis-git** [comment]
Still actual!

**Comment by yiliansource** [comment]
@slorber I'm afraid this is a little above me, I managed to get the generation working perfectly on my local machine, but deployment (even just for development testing on Vercel) has too many intricacies and limitations for me to handle.

**Comment by slorber** [comment]
NP, I'll handle that later.

Afaik for Vercel free plan we have to reduce memory limit to 1024 or something, had to do this conf in the past:
https://github.com/slorber/og-image-quotes/blob/master/now.json

**Comment by slorber** [comment]
For later usage: GitHub has published a blog post on how they generate their social cards: 

https://github.blog/2021-06-22-framework-building-open-graph-images/

**Comment by PatelN123** [comment]
Hey @slorber,

I managed to get a working instance of this social cards feature working on my own site. I used the vercel og image tool. It works a charm. 
A good idea for this as @Josh-Cena was telling me about, is to have a config option to set your URL of your own vercel og image instance or even as you mentioned here calling it docusaurus-image and docusaurus can add that. 

A way I have done it is by adding this file: [here](https://github.com/PatelN123/Digital-Support-Notes/blob/main/Website/src/theme/Seo/index.js)

What are your thoughts on this design?

**Comment by slorber** [comment]
@PatelN123 ideally you should not need to use swizzle on the Seo component to provide the social cards feature

Also, we want it to be flexible, so ideally the user should be able to provide a callback directly in the config to generate the social card URL, adapting to any card service (not just the one we suggest). 

Eventually, we could even provide a lightweight package that builds social cards as part of the build process locally

**Comment by dpang314** [comment]
I would like to work on this. 

I was able to get passing in a function working, but only inside of the main docusaurus.config.js as opposed inside the theme config. Inside of the theme config, the function is evaluated to a string rather than returning a callback. I think that this is because theme data (must be serialized to plain strings)[https://docusaurus.io/docs/advanced/plugins#theme-design]. 

Is it fine to leave the function inside the main config?

**Comment by Josh-Cena** [comment]
@dpang314 yes, if it's actually accessed on server-side (and I hope it is!), you can put it in main config... for now. The actual API design can be deferred. A working prototype is fine.

**Comment by dpang314** [comment]
I have a working proof of concept that is deployed here: https://capable-belekoy-059ce2.netlify.app/

Currently I haven’t fully implemented it, but it should work for all docs. I don’t think it will be difficult to apply the same implementation to pages and blogs. 

A few examples:
[Page 1](https://capable-belekoy-059ce2.netlify.app/docs/next/api/plugins/@docusaurus/plugin-content-docs) | [Preview 1](https://dnschecker.org/open-graph-preview-generate-metatags.php?query=https%3A%2F%2Fcapable-belekoy-059ce2.netlify.app%2Fdocs%2Fnext%2Fapi%2Fplugins%2F%40docusaurus%2Fplugin-content-docs) 
[Page 2](https://capable-belekoy-059ce2.netlify.app/docs/installation) | [Preview 2](https://dnschecker.org/open-graph-preview-generate-metatags.php?query=https%3A%2F%2Fcapable-belekoy-059ce2.netlify.app%2Fdocs%2Finstallation)
[Page 3](https://capable-belekoy-059ce2.netlify.app/docs/2.0.0-beta.20/playground) | [Preview 3](https://dnschecker.org/open-graph-preview-generate-metatags.php?query=https%3A%2F%2Fcapable-belekoy-059ce2.netlify.app%2Fdocs%2F2.0.0-beta.20%2Fplayground)
[Category](https://capable-belekoy-059ce2.netlify.app/docs/category/getting-started) | [Preview](https://dnschecker.org/open-graph-preview-generate-metatags.php?query=https%3A%2F%2Fcapable-belekoy-059ce2.netlify.app%2Fdocs%2Fcategory%2Fgetting-started)

In the implementation, the docs, blog, and pages plugins would save the generated url for each route to global data. During server side generation, for each route, I check the global data to get the url for that route (and use a default if none is found).

This implementation keeps everything on server side and centralizes the process. This makes things easier because as far as I can tell, I can't pass the function to the frontend. 

Will this implementation be feasible?

**Comment by slorber** [comment]
@dpang314 I'd like to see the code to be able to tell :)

What matters most is to have a good/flexible public API. 

If it's not possible to find the best API today, we can still ship this as an experimental version and do breaking changes later if we have a better approach.

**Comment by dpang314** [comment]
Link to the [code](https://github.com/dpang314/docusaurus/tree/social-card-proof-of-concept). I have a lot to clean up and add, but the basic implementation is there.

**Comment by slorber** [comment]
@rotemtam implemended social cards in userland with swizzle:

https://github.com/ariga/atlas/pull/1201
https://twitter.com/_rtam/status/1585301256113098752

I hope we can find a good API for this feature, in the meantime it is definitively possible and quite easy to implement in userland ;)

**Comment by anaclumos** [comment]
Linking this for reference: https://github.com/vercel/satori

**Comment by MChetanOP** [comment]
anybody have video tutorial i don't [really] understand im noob , Tell me how to create Square image preview for facebook 

Check here the sample dont know how it works - [click ](https://www.facebook.com/194473201285542/posts/pfbid08BrkLWXrqJ3Vhy4moZrbgaC4p2TzxXh2L74bgZs2u6gH1kn5aZ2Bgcs2A4SKDQPPl/?sfnsn=wiwspmo&mibextid=6aamW6)

**Comment by nandorojo** [comment]
Has anyone gotten satori working with Docusaurus?

**Comment by anaclumos** [comment]
> Has anyone gotten Satori working with Docusaurus?

@nandorojo I have. https://github.com/anaclumos/extracranial/blob/74b2d0be330136b95b595e818abdbb1c3f1d9183/src/theme/DocItem/Metadata/index.tsx#L18-L20

The Satori instance itself is at https://github.com/anaclumos/og

**Comment by Zenahr** [comment]
I tried adding a static social card but the option "image" is unrecognized by docusaurus v^2.0 and I can't seem to find any documentation about adding (global) social cards here: https://docusaurus.io/docs/seo

trace:

```js
[ERROR] Error: These field(s) ("image",) are not recognized in docusaurus.config.js.
If you still want these fields to be in your configuration, put them in the "customFields" field.
See https://docusaurus.io/docs/api/docusaurus-config/#customfields    
    at validateConfig (C:\PRODUCT-DOCS\node_modules\@docusaurus\core\lib\server\configValidation.js:228:15)
    at loadSiteConfig (C:\PRODUCT-DOCS\node_modules\@docusaurus\core\lib\server\config.js:40:62)
    at async loadContext (C:\PRODUCT-DOCS\node_modules\@docusaurus\core\lib\server\index.js:31:63)
    at async load (C:\PRODUCT-DOCS\node_modules\@docusaurus\core\lib\server\index.js:74:21)
    at async Command.start (C:\PRODUCT-DOCS\node_modules\@docusaurus\core\lib\commands\start.js:44:19)
[INFO] Docusaurus version: 2.4.1
Node version: v18.12.1
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

```

**Comment by kamranayub** [comment]
I'll be investigating this for our work over on excaliburjs/Excalibur and we are on CF Pages; which means I think we could bake this into a Cloudflare Pages Function to handle (and they have a generous free tier + we can take advantage of CDN caching). As another alternative to Vercel. I haven't looked over it in detail, I've only skimmed over the thread.

**Comment by jcubic** [comment]
Just my few cents, since I found this issue.

How about using JSX like everything in Docusaurs and create SVG file out of it. Then render it into pure SVG file that is then converted to png. On Build time but only on `npm run build`.

**Comment by jcubic** [comment]
Sure, but the API can be literally the same, it only needs to be triggered on every page instead of only blog posts, and you need to pick a good name.

I also find all missing properties except `slug`. And still have no idea how to add this to HTML.

**Comment by slorber** [comment]
> Sure, but the API can be literally the same, it only needs

@jcubic "it only" assumes it's easy, but from what I read you don't plainly understand the problem space, and only see your use case while we try to generalize a solution. 

This feature is not related to the blog, not related to the feed. Our other plugins and third-party plugins might as well have a behavior that is far from how the blog plugin works, so don't think that a solution for a blog could be generalized that easily. Unlike blog posts, docs have versions. It is not just a "list of items" on which there is an image attribute to set.

We already have a [`createFrontMatter`](https://docusaurus.io/blog/releases/3.1#parsefrontmatter-hook
) generic API that you could use to automatically generate an extra `image` front matter if you want to. But that will only work for Markdown-centric documents, and not React pages. This might be good enough for your use case, but is definitively not the general solution we try to create here.

We produce pages for which we might want to generate a custom social card, that do not even belong to a Markdown document or React page component. This is notably the case for blog tags pages, blog paginated lists, blog archive page, upcoming blog authors pages, docs autogenerated categories, and more fancy pages that we could create in the future.

**Comment by armchairancap** [comment]
https://github.com/Kas-tle/docusaurus-og with Vercel Satori. Haven't tried it.
